src/lib/io/sol-memmap-storage.c

```diff
@@ -489,20 +708,41 @@ sol_memmap_add_map(const struct sol_memmap_map *map)
         return -EINVAL;
     }
 
+    timeout = sol_vector_append(&write_timeouts);
+    SOL_NULL_CHECK(timeout, -ENOMEM);
+
+    timeout->map = resolved_map;
+    sol_vector_init(&timeout->pending_writes, sizeof(struct pending_write_data));
+
     r = sol_ptr_vector_append(&memory_maps, (void *)resolved_map);
-    SOL_INT_CHECK(r, < 0, r);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
 
     return 0;
+
+error:
+    sol_vector_del(&write_timeouts, write_timeouts.len - 1);
+    return r;
 }
 
 SOL_API int
 sol_memmap_remove_map(const struct sol_memmap_map *map)
 {
     struct sol_memmap_map *iter;
+    struct map_timeout *map_timeout;
     int i;
 
     SOL_NULL_CHECK(map, -EINVAL);
 
+    SOL_VECTOR_FOREACH_IDX (&write_timeouts, map_timeout, i) {
+        if (map_timeout->map->entries == map->entries) {
+            if (map_timeout->timeout) {
+                perform_pending_writes((void *)map_timeout->map);
+                sol_timeout_del(map_timeout->timeout);
```

Comment:
you should be doing the `sol_timeout_del()` before calling `perform_pending_writes()` as that function will make the pointer to `struct sol_timeout` NULL (or keep a copy of the pointer in your stack).